File: bindings/cpp/src/ed25519.cpp
// ed25519.cpp - RFC 8032 over the magnitude bignum layer.

#include "ed25519.hpp"

#include <mutex>
#include <optional>
#include <stdexcept>

#include "bignum.hpp"
#include "sha2.hpp"

namespace co {
namespace ed25519 {

namespace {

// The curve constants, computed once on first use.
struct Curve {
    Big p;        // 2^255 - 19
    Big q;        // the group order (2^252 + 27742317...)
    Big d;        // -121665 / 121666 mod p
    Big sqrt_m1;  // 2^((p-1)/4) mod p
    Big p_minus_2;
    Big p_plus_3_div_8;
};

struct Pt {
    Big x, y, z, t;  // extended twisted Edwards coordinates
};

Curve* g_curve = nullptr;
Pt* g_G = nullptr;
std::once_flag g_once;

// ---- field arithmetic mod p ---------------------------------------------

// Fast reduction mod 2^255 - 19: fold the high part times 19 into the low.
Big mod_p(Big a) {
    const Big& p = g_curve->p;
    const Big nineteen = big_from_u64(19);
    while (big_bitlen(a) > 255) {
        Big hi = big_shr(a, 255);
        Big lo = big_lowbits(a, 255);
        a = big_add(lo, big_mul(hi, nineteen));
    }
    while (big_cmp(a, p) >= 0) a = big_sub(a, p);
    return a;
}

Big fadd(const Big& a, const Big& b) { return mod_p(big_add(a, b)); }

// a - b mod p with a, b already reduced: computed as a + p - b (the
// magnitude layer has no negatives; this mirrors Python's floored %).
Big fsub(const Big& a, const Big& b) {
    return mod_p(big_sub(big_add(a, g_curve->p), b));
}

Big fmul(const Big& a, const Big& b) { return mod_p(big_mul(a, b)); }

// Square-and-multiply, reducing with the fast p fold at every step.
Big modpow_p(const Big& b, const Big& e) {
    Big result = big_from_u64(1);
    Big base = mod_p(b);
    for (size_t i = big_bitlen(e); i-- > 0;) {
        result = fmul(result, result);
        if (big_bit(e, i)) result = fmul(result, base);
    }
    return result;
}

// Fermat inversion: x^(p-2) mod p.
Big inv_p(const Big& x) { return modpow_p(x, g_curve->p_minus_2); }

// ---- the point group ------------------------------------------------------

Pt point_add(const Pt& P, const Pt& Q) {
    const Big two = big_from_u64(2);
    Big A = fmul(fsub(P.y, P.x), fsub(Q.y, Q.x));
    Big B = fmul(fadd(P.y, P.x), fadd(Q.y, Q.x));
    Big C = fmul(fmul(fmul(two, P.t), Q.t), g_curve->d);
    Big D = fmul(fmul(two, P.z), Q.z);
    Big E = fsub(B, A);
    Big F = fsub(D, C);
    Big G = fadd(D, C);
    Big H = fadd(B, A);
    return Pt{fmul(E, F), fmul(G, H), fmul(F, G), fmul(E, H)};
}

Pt point_mul(const Big& s, Pt P) {
    // The neutral element (0, 1, 1, 0).
    Pt Q{Big{}, big_from_u64(1), big_from_u64(1), Big{}};
    size_t bits = big_bitlen(s);
    for (size_t i = 0; i < bits; ++i) {
        if (big_bit(s, i)) Q = point_add(Q, P);
        P = point_add(P, P);
    }
    return Q;
}

bool point_equal(const Pt& P, const Pt& Q) {
    // Cross-multiplied projective comparison, kept non-negative.
    if (big_cmp(fmul(P.x, Q.z), fmul(Q.x, P.z)) != 0) return false;
    if (big_cmp(fmul(P.y, Q.z), fmul(Q.y, P.z)) != 0) return false;
    return true;
}

std::optional<Big> recover_x(const Big& y, bool sign) {
    if (big_cmp(y, g_curve->p) >= 0) return std::nullopt;
    const Big one = big_from_u64(1);
    Big y2 = fmul(y, y);
    Big x2 = fmul(fsub(y2, one), inv_p(fadd(fmul(g_curve->d, y2), one)));
    if (x2.isZero()) {
        if (sign) return std::nullopt;
        return Big{};
    }
    Big x = modpow_p(x2, g_curve->p_plus_3_div_8);
    if (!fsub(fmul(x, x), x2).isZero()) x = fmul(x, g_curve->sqrt_m1);
    if (!fsub(fmul(x, x), x2).isZero()) return std::nullopt;
    if (big_bit(x, 0) != sign) x = big_sub(g_curve->p, x);
    return x;
}

std::string point_compress(const Pt& P) {
    Big zinv = inv_p(P.z);
    Big x = fmul(P.x, zinv);
    Big y = fmul(P.y, zinv);
    std::string out = big_to_bytes_le(y, 32);
    if (big_bit(x, 0))
        out[31] = static_cast<char>(static_cast<unsigned char>(out[31]) | 0x80);
    return out;
}

std::optional<Pt> point_decompress(const std::string& s) {
    if (s.size() != 32) return std::nullopt;
    Big raw = big_from_bytes_le(s);
    bool sign = big_bit(raw, 255);
    Big y = big_lowbits(raw, 255);
    std::optional<Big> x = recover_x(y, sign);
    if (!x) return std::nullopt;
    return Pt{*x, y, big_from_u64(1), fmul(*x, y)};
}

// ---- setup ---------------------------------------------------------------

void init_curve() {
    g_curve = new Curve();
    // p = 2^255 - 19.
    g_curve->p = big_sub(big_shl(big_from_u64(1), 255), big_from_u64(19));
    // q = 2^252 + 27742317777372353535851937790883648493.
    g_curve->q = big_from_hex(
        "1000000000000000000000000000000014def9dea2f79cd65812631a5cf5d3ed");
    g_curve->p_minus_2 = big_sub(g_curve->p, big_from_u64(2));
    g_curve->p_plus_3_div_8 = big_shr(big_add(g_curve->p, big_from_u64(3)), 3);
    // d = -121665 * inv(121666) mod p, kept non-negative as (p - 121665)/121666.
    g_curve->d = fmul(big_sub(g_curve->p, big_from_u64(121665)),
                      inv_p(big_from_u64(121666)));
    // sqrt(-1) = 2^((p-1)/4) mod p.
    g_curve->sqrt_m1 = modpow_p(
        big_from_u64(2), big_shr(big_sub(g_curve->p, big_from_u64(1)), 2));
    // The base point: y = 4/5, x recovered with the even sign.
    Big g_y = fmul(big_from_u64(4), inv_p(big_from_u64(5)));
    std::optional<Big> g_x = recover_x(g_y, false);
    if (!g_x) throw std::runtime_error("ed25519: base point recovery failed");
    g_G = new Pt{*g_x, g_y, big_from_u64(1), fmul(*g_x, g_y)};
}

void ensure_init() { std::call_once(g_once, init_curve); }

// ---- scalars and hashing --------------------------------------------------

// (a, prefix): the clamped secret scalar and the 32-byte nonce prefix.
std::pair<Big, std::string> secret_expand(const std::string& secret) {
    if (secret.size() != 32)
        throw std::runtime_error("secret key must be 32 bytes");
    std::string h = sha512(secret);
    std::string scalar = h.substr(0, 32);
    // The RFC 8032 clamp: clear bits 0-2 and bit 255, set bit 254.
    scalar[0] = static_cast<char>(static_cast<unsigned char>(scalar[0]) & 248);
    scalar[31] = static_cast<char>(static_cast<unsigned char>(scalar[31]) & 127);
    scalar[31] = static_cast<char>(static_cast<unsigned char>(scalar[31]) | 64);
    return {big_from_bytes_le(scalar), h.substr(32)};
}

Big sha512_modq(const std::string& s) {
    return big_mod(big_from_bytes_le(sha512(s)), g_curve->q);
}

}  // namespace

std::string secret_to_public(const std::string& secret) {
    ensure_init();
    auto [a, prefix] = secret_expand(secret);
    (void)prefix;
    return point_compress(point_mul(a, *g_G));
}

std::string sign(const std::string& secret, const std::string& msg) {
    ensure_init();
    auto [a, prefix] = secret_expand(secret);
    std::string A = point_compress(point_mul(a, *g_G));
    Big r = sha512_modq(prefix + msg);
    std::string Rs = point_compress(point_mul(r, *g_G));
    Big h = sha512_modq(Rs + A + msg);
    Big s = big_mod(big_add(r, big_mul(h, a)), g_curve->q);
    return Rs + big_to_bytes_le(s, 32);
}

bool verify(const std::string& public_key, const std::string& msg,
            const std::string& signature) {
    ensure_init();
    if (public_key.size() != 32 || signature.size() != 64) return false;
    std::optional<Pt> A = point_decompress(public_key);
    if (!A) return false;
    std::string Rs = signature.substr(0, 32);
    std::optional<Pt> R = point_decompress(Rs);
    if (!R) return false;
    Big s = big_from_bytes_le(signature.substr(32));
    if (big_cmp(s, g_curve->q) >= 0) return false;
    Big h = sha512_modq(Rs + public_key + msg);
    Pt sB = point_mul(s, *g_G);
    Pt hA = point_mul(h, *A);
    return point_equal(sB, point_add(*R, hA));
}

}  // namespace ed25519
}  // namespace co

File: bindings/kotlin/src/Ed25519.kt
// Ed25519 digital signatures (RFC 8032), pure Kotlin, ported line-for-line
// from the Python binding's ed25519.py over the Bignum layer.
//
// Slow but correct: intended for the conformance suite and small tools.
// All field arithmetic stays non-negative (a - b mod p is computed as
// a + p - b), and reduction mod p = 2^255 - 19 uses the limb-aligned fold
// 2^256 = 38 (mod p). Gated on the RFC 8032 TEST 1 known answer by
// checkKnownAnswer() (called from the conformance runner's internal checks).
package org.causalontology

object Ed25519 {

    // p = 2^255 - 19 (the field prime).
    private val P = Bignum.fromHex(
        "7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffed")
    // q = 2^252 + 27742317777372353535851937790883648493 (the group order).
    private val Q = Bignum.fromHex(
        "1000000000000000000000000000000014def9dea2f79cd65812631a5cf5d3ed")

    private val THIRTY_EIGHT = intArrayOf(38)

    // ------------------------------------------------------ field arithmetic
    // Reduce an arbitrary product to [0, p) by folding at the 16-limb (2^256)
    // boundary: x = hi * 2^256 + lo = hi * 38 + lo (mod p).
    private fun reduceP(xIn: IntArray): IntArray {
        var x = xIn
        while (x.size > 16) {
            val lo = Bignum.norm(x.copyOf(16))
            val hi = x.copyOfRange(16, x.size)
            x = Bignum.add(lo, Bignum.mul(hi, THIRTY_EIGHT))
        }
        while (Bignum.cmp(x, P) >= 0) x = Bignum.sub(x, P)
        return x
    }

    private fun addP(a: IntArray, b: IntArray): IntArray = reduceP(Bignum.add(a, b))

    // a - b mod p, computed non-negatively as (a + p) - b.
    private fun subP(a: IntArray, b: IntArray): IntArray =
        reduceP(Bignum.sub(Bignum.add(a, P), b))

    private fun mulP(a: IntArray, b: IntArray): IntArray = reduceP(Bignum.mul(a, b))

    // Modular exponentiation mod p by square-and-multiply over exp's bits.
    private fun powP(base: IntArray, exp: IntArray): IntArray {
        var result = Bignum.ONE
        var b = reduceP(base)
        val bits = Bignum.bitLen(exp)
        for (i in 0 until bits) {
            if (Bignum.testBit(exp, i)) result = mulP(result, b)
            b = mulP(b, b)
        }
        return result
    }

    // Modular inverse by Fermat's little theorem: x^(p-2) mod p.
    private fun invP(x: IntArray): IntArray = powP(x, Bignum.sub(P, Bignum.fromLong(2)))

    // ------------------------------------------------------- curve constants
    // d = -121665 * inv(121666) mod p.
    private val D = mulP(
        Bignum.sub(P, Bignum.fromLong(121665)),
        invP(Bignum.fromLong(121666)))

    // sqrt(-1) = 2^((p-1)/4) mod p.
    private val SQRT_M1 = powP(
        Bignum.fromLong(2),
        Bignum.shr(Bignum.sub(P, Bignum.ONE), 2))

    // (p + 3) / 8, the exponent used in x-recovery.
    private val PP3_D8 = Bignum.shr(Bignum.add(P, Bignum.fromLong(3)), 3)

    // A point is (X, Y, Z, T) in extended homogeneous coordinates.
    private class Point(val x: IntArray, val y: IntArray, val z: IntArray, val t: IntArray)

    // The neutral element (0, 1, 1, 0).
    private val NEUTRAL = Point(Bignum.ZERO, Bignum.ONE, Bignum.ONE, Bignum.ZERO)

    // The base point G: y = 4 * inv(5) mod p, x recovered with sign 0.
    private val G: Point = run {
        val gy = mulP(Bignum.fromLong(4), invP(Bignum.fromLong(5)))
        val gx = recoverX(gy, 0) ?: throw IllegalStateException("base point recovery failed")
        Point(gx, gy, Bignum.ONE, mulP(gx, gy))
    }

    // ------------------------------------------------------------ point ops
    private fun pointAdd(p1: Point, p2: Point): Point {
        val a = mulP(subP(p1.y, p1.x), subP(p2.y, p2.x))
        val b = mulP(addP(p1.y, p1.x), addP(p2.y, p2.x))
        val c = mulP(mulP(Bignum.fromLong(2), mulP(p1.t, p2.t)), D)
        val d2 = mulP(Bignum.fromLong(2), mulP(p1.z, p2.z))
        val e = subP(b, a)
        val f = subP(d2, c)
        val g = addP(d2, c)
        val h = addP(b, a)
        return Point(mulP(e, f), mulP(g, h), mulP(f, g), mulP(e, h))
    }

    private fun pointMul(sIn: IntArray, pIn: Point): Point {
        var q = NEUTRAL
        var p1 = pIn
        val bits = Bignum.bitLen(sIn)
        for (i in 0 until bits) {
            if (Bignum.testBit(sIn, i)) q = pointAdd(q, p1)
            p1 = pointAdd(p1, p1)
        }
        return q
    }

    private fun pointEqual(p1: Point, p2: Point): Boolean {
        // Cross-multiplied comparison avoids inversions; both sides are in [0, p).
        if (Bignum.cmp(mulP(p1.x, p2.z), mulP(p2.x, p1.z)) != 0) return false
        if (Bignum.cmp(mulP(p1.y, p2.z), mulP(p2.y, p1.z)) != 0) return false
        return true
    }

    // ------------------------------------------------- compression / recovery
    private fun recoverX(y: IntArray, sign: Int): IntArray? {
        if (Bignum.cmp(y, P) >= 0) return null
        val y2 = mulP(y, y)
        val x2 = mulP(subP(y2, Bignum.ONE), invP(addP(mulP(D, y2), Bignum.ONE)))
        if (Bignum.isZero(x2)) {
            return if (sign != 0) null else Bignum.ZERO
        }
        var x = powP(x2, PP3_D8)
        if (!Bignum.isZero(subP(mulP(x, x), x2))) {
            x = mulP(x, SQRT_M1)
        }
        if (!Bignum.isZero(subP(mulP(x, x), x2))) return null
        if ((if (Bignum.testBit(x, 0)) 1 else 0) != sign) {
            x = Bignum.sub(P, x)
        }
        return x
    }

    private fun pointCompress(p1: Point): ByteArray {
        val zinv = invP(p1.z)
        val x = mulP(p1.x, zinv)
        val y = mulP(p1.y, zinv)
        val out = Bignum.toBytesLE(y, 32)
        if (Bignum.testBit(x, 0)) {
            out[31] = (out[31].toInt() or 0x80).toByte()
        }
        return out
    }

    private fun pointDecompress(s: ByteArray): Point? {
        if (s.size != 32) return null
        val sign = (s[31].toInt() ushr 7) and 1
        val yBytes = s.copyOf()
        yBytes[31] = (yBytes[31].toInt() and 0x7F).toByte()
        val y = Bignum.fromBytesLE(yBytes)
        val x = recoverX(y, sign) ?: return null
        return Point(x, y, Bignum.ONE, mulP(x, y))
    }

    // ------------------------------------------------------ scalars and keys
    private fun secretExpand(secret: ByteArray): Pair<IntArray, ByteArray> {
        if (secret.size != 32) throw IllegalArgumentException("secret key must be 32 bytes")
        val h = Sha2.sha512(secret)
        val a = h.copyOfRange(0, 32)
        // Clamp: clear bits 0..2 and bit 255, set bit 254 (RFC 8032).
        a[0] = (a[0].toInt() and 0xF8).toByte()
        a[31] = (a[31].toInt() and 0x3F).toByte()
        a[31] = (a[31].toInt() or 0x40).toByte()
        return Pair(Bignum.fromBytesLE(a), h.copyOfRange(32, 64))
    }

    private fun sha512ModQ(msg: ByteArray): IntArray =
        Bignum.mod(Bignum.fromBytesLE(Sha2.sha512(msg)), Q)

    // ------------------------------------------------------------ public API
    fun secretToPublic(secret: ByteArray): ByteArray {
        val (a, _) = secretExpand(secret)
        return pointCompress(pointMul(a, G))
    }

    fun sign(secret: ByteArray, msg: ByteArray): ByteArray {
        val (a, prefix) = secretExpand(secret)
        val aPub = pointCompress(pointMul(a, G))
        val r = sha512ModQ(prefix + msg)
        val rs = pointCompress(pointMul(r, G))
        val h = sha512ModQ(rs + aPub + msg)
        val s = Bignum.mod(Bignum.add(r, Bignum.mul(h, a)), Q)
        return rs + Bignum.toBytesLE(s, 32)
    }

    fun verify(public: ByteArray, msg: ByteArray, signature: ByteArray): Boolean {
        if (public.size != 32 || signature.size != 64) return false
        val aPoint = pointDecompress(public) ?: return false
        val rs = signature.copyOfRange(0, 32)
        val rPoint = pointDecompress(rs) ?: return false
        val s = Bignum.fromBytesLE(signature.copyOfRange(32, 64))
        if (Bignum.cmp(s, Q) >= 0) return false
        val h = sha512ModQ(rs + public + msg)
        val sB = pointMul(s, G)
        val hA = pointMul(h, aPoint)
        return pointEqual(sB, pointAdd(rPoint, hA))
    }

    // --------------------------------------------------- known-answer gate
    // RFC 8032 section 7.1 TEST 1: seed, public key, and the exact signature
    // of the empty message; plus rejection of a wrong message.
    fun checkKnownAnswer() {
        val seed = hexToBytes(
            "9d61b19deffd5a60ba844af492ec2cc44449c5697b326919703bac031cae7f60")!!
        val pub = secretToPublic(seed)
        check(bytesToHex(pub) ==
            "d75a980182b10ab7d54bfed3c964073a0ee172f3daa62325af021a68f707511a") {
            "RFC 8032 TEST 1 public key mismatch: ${bytesToHex(pub)}"
        }
        val sig = sign(seed, ByteArray(0))
        check(bytesToHex(sig) ==
            "e5564300c360ac729086e2cc806e828a84877f1eb8e5d974d873e06522490155" +
            "5fb8821590a33bacc61e39701cf9b46bd25bf5f0595bbe24655141438e7a100b") {
            "RFC 8032 TEST 1 signature mismatch: ${bytesToHex(sig)}"
        }
        check(verify(pub, ByteArray(0), sig)) { "RFC 8032 TEST 1 signature must verify" }
        check(!verify(pub, "x".encodeToByteArray(), sig)) {
            "RFC 8032 TEST 1 signature must not verify a different message"
        }
    }
}
